File: Zamba/Zamba.Web/Zamba.FUS/ZOptBusiness.cs

using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using Zamba.AppBlock;

public class ZOptBusiness
{
    public static void Insert(string key, string value)
    {
        if (Zamba.Core.Cache.UsersAndGroups.hsOptions.ContainsKey(key) == false)
        {
            Zamba.Core.Cache.UsersAndGroups.hsOptions.Add(key, value);
        }
        else
        {
            Zamba.Core.Cache.UsersAndGroups.hsOptions[key] = value;
        }
        ZOptFactory.Insert(key, value);
    }

    public static void Update(string key, string value)
    {
        if (Zamba.Core.Cache.UsersAndGroups.hsOptions.ContainsKey(key))
        {
            Zamba.Core.Cache.UsersAndGroups.hsOptions[key] = value;
        }
        ZOptFactory.Update(key, value);
    }

    public static string GetValue(string key)
    {
        if (Zamba.Core.Cache.UsersAndGroups.hsOptions.ContainsKey(key))
        {
            if (Zamba.Core.Cache.UsersAndGroups.hsOptions[key] != null)
                return Zamba.Core.Cache.UsersAndGroups.hsOptions[key].ToString();
            else
                return string.Empty;
        }
        string Value = ZOptFactory.GetValue(key);
        if (Value == null)
        {
            return string.Empty;
        }

        if (Zamba.Core.Cache.UsersAndGroups.hsOptions.ContainsKey(key) == false)
            Zamba.Core.Cache.UsersAndGroups.hsOptions.Add(key, Value);
        return Value;
    }

    public static string GetValueOrDefault(string key, string DefaultValue)
    {
        if (Zamba.Core.Cache.UsersAndGroups.hsOptions.ContainsKey(key))
        {
            if (Zamba.Core.Cache.UsersAndGroups.hsOptions[key] != null)
                return Zamba.Core.Cache.UsersAndGroups.hsOptions[key].ToString();
            else
                return string.Empty;
        }
        string Value = ZOptFactory.GetValue(key);
        if (Value == null || Value.Length == 0)
        {
            Insert(key, DefaultValue);
            Value = DefaultValue;
        }
        if (Zamba.Core.Cache.UsersAndGroups.hsOptions.ContainsKey(key) == false)
            Zamba.Core.Cache.UsersAndGroups.hsOptions.Add(key, Value);
        return Value;
    }

    public static void GetAllValues()
    {

        DataSet values = null;
        try
        {
            values = ZOptFactory.GetValues();
            lock (Zamba.Core.Cache.UsersAndGroups.hsOptions)
            {
                foreach (DataRow Value in values.Tables[0].Rows)
                {
                    if (Zamba.Core.Cache.UsersAndGroups.hsOptions.ContainsKey(Value[1]) == false)
                        Zamba.Core.Cache.UsersAndGroups.hsOptions.Add(Value[1], Value[0]);
                }
            }
        }
        catch (Exception ex)
        {
            ZException.Log(new Exception("Error al cargar las preferencias de maquina" + ex.ToString()));
        }
        finally
        {
            if (values != null)
            {
                values.Dispose();
                values = null;
            }
        }




    }

    public static void InsertUpdateValue(string key, string value)
    {
        if (Zamba.Core.Cache.UsersAndGroups.hsOptions.ContainsKey(key) == false)
        {
            Zamba.Core.Cache.UsersAndGroups.hsOptions.Add(key, value);
        }
        else
        {
            Zamba.Core.Cache.UsersAndGroups.hsOptions[key] = value;
        }
        ZOptFactory.InsertUpdateValue(key, value);
    }

}

File: Zamba/ZambaWeb2013/zamba2010/Zamba.Bussines/ZOptBusiness.vb
Imports Zamba.Data

Public Class ZOptBusiness

    Public Shared Sub Insert(ByVal key As String, ByVal value As String)
        ZOptFactory.Insert(key, value)
    End Sub

    Public Shared Sub Update(ByVal key As String, ByVal value As String)
        ZOptFactory.Update(key, value)
    End Sub

    Public Function GetValue(ByVal key As String) As String
        If Cache.UsersAndGroups.hsOptions.ContainsKey(key) Then
            Return Cache.UsersAndGroups.hsOptions(key)
        End If

        Dim zoptf As New ZOptFactory
        Dim Value As String = zoptf.GetValue(key)
        zoptf = Nothing

        '  If Value Is Nothing Then
        'ZTrace.WriteLineIf(ZTrace.IsError, "El valor de la clave " & key & _
        '                 " no se encuentra configurada en ZOPT. Esto puede ocasionar errores o comportamientos no esperados en la aplicación.")
        ' End If

        SyncLock Cache.UsersAndGroups.hsOptions.SyncRoot
            If Not Cache.UsersAndGroups.hsOptions.Contains(key) Then
                Cache.UsersAndGroups.hsOptions.Add(key, Value)
            End If
        End SyncLock

        Return Value
    End Function

    Public Shared Function GetValueOrDefault(ByVal key As String, DefaultValue As String) As String
        If Cache.UsersAndGroups.hsOptions.ContainsKey(key) Then
            Return Cache.UsersAndGroups.hsOptions(key)
        End If

        Dim zoptf As New ZOptFactory
        Dim Value As String = zoptf.GetValue(key)
        zoptf = Nothing

        If Value Is Nothing Then
            InsertUpdateValue(key, DefaultValue)
            Value = DefaultValue
        End If
        If Cache.UsersAndGroups.hsOptions.ContainsKey(key) = False Then Cache.UsersAndGroups.hsOptions.Add(key, Value)
        Return Value
    End Function

    Public Function GetValueOrDefaultNonShared(ByVal key As String, DefaultValue As String) As String
        If Cache.UsersAndGroups.hsOptions.ContainsKey(key) Then
            Return Cache.UsersAndGroups.hsOptions(key)
        End If

        Dim zoptf As New ZOptFactory
        Dim Value As String = zoptf.GetValue(key)
        zoptf = Nothing

        If Value Is Nothing AndAlso DefaultValue <> String.Empty Then
            InsertUpdateValue(key, DefaultValue)
            Value = DefaultValue
        End If
        If Cache.UsersAndGroups.hsOptions.ContainsKey(key) = False Then Cache.UsersAndGroups.hsOptions.Add(key, Value)
        Return Value
    End Function

    Public Sub LoadAllOptions()
        If Cache.UsersAndGroups.hsOptions.Count = 0 Then
            SyncLock Cache.UsersAndGroups.hsOptions
                Dim values As DataTable = Nothing
                Try
                    values = ZOptFactory.getAllOptionsValues()
                    For Each Value As DataRow In values.Rows
                        If Cache.UsersAndGroups.hsOptions.ContainsKey(Value(0)) = False Then Cache.UsersAndGroups.hsOptions.Add(Value(0), Value(1))
                    Next
                Catch ex As Exception
                    Zamba.Core.ZClass.raiseerror(New Exception("Error al cargar las preferencias de maquina. " & ex.ToString))
                Finally
                    If values IsNot Nothing Then
                        values.Dispose()
                        values = Nothing
                    End If
                End Try
            End SyncLock
        End If
    End Sub

    Public Shared Sub InsertUpdateValue(ByVal key As String, ByVal value As String)
        If Cache.UsersAndGroups.hsOptions.ContainsKey(key) = False Then
            Cache.UsersAndGroups.hsOptions.Add(key, value)
        Else
            Cache.UsersAndGroups.hsOptions(key) = value
        End If

        Dim zoptf As New ZOptFactory
        zoptf.InsertUpdateValue(key, value)
        zoptf = Nothing
    End Sub

    Public Shared Sub ClearCache()
        Cache.UsersAndGroups.hsOptions = New SynchronizedHashtable
    End Sub
End Class
